net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+	/*
+	* Deterministically combine portal and tmid into a unique 16-bit port
+	* number (greater than 1024). Tricky.
+	*
+	* Port number is, in binary: tttttttttt1ppppp, that is, 10 bits of tmid
+	* (which must be less than 1024), followed by a set bit (guaranteeing
+	* that the port is not reserved), followed by 5 bits of (portal - 30),
+	* so that portal must be in the range 30..61.
+	*
+	if (tmid >= 1024 || (portal - 30) >= 32)
+		return M0_ERR_INFO(-EPROTO,
+			"portal: %u, tmid: %u", portal, tmid);
+	*/
+
+	if ( portal < 30)
+		portal = 30 + portal;
+
+	portnum  = tmid | (1 << 10) | ((portal - 30) << 11);
+	M0_ASSERT(portnum < 65537);
+	sprintf(port, "%d", portnum);
+	fab_autotm[tmid] = 1;
+	return M0_RC(0);
+}
+
+/**
+ * This function decodes the socket format address and extracts the ip address
+ * and port number from it.
+ */
+static int libfab_ep_addr_decode_sock(const char *ep_name, char *node,
+				      size_t nodeSize, char *port,
+				      size_t portSize)
+{
+	int   shift = 0;
+	int   f;
+	int   s;
+	char *at;
+
+	for (f = 0; f < ARRAY_SIZE(portf) ; ++f) {
+		if (portf[f]!= NULL) {
+			shift = strlen(portf[f]);
+			if (strncmp(ep_name, portf[f], shift) == 0)
+				break;
+		}
+	}
+	if (ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	for (s = 0; s < ARRAY_SIZE(socktype); ++s) {
+		if (socktype[s] != NULL) {
+			shift = strlen(socktype[s]);
+			if (strncmp(ep_name, socktype[s], shift) == 0)
+				break;
+		}
+	}
+	if (ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	at = strchr(ep_name, '@');
+	if (at == NULL) {
+		return M0_ERR(-EINVAL);
+	} else {
+		at++;
+		if (at == NULL)
+			return M0_ERR(-EINVAL);
+		M0_PRE(portSize >= (strlen(at)+1));
+		memcpy(port,at,(strlen(at)+1));
+	}
+	M0_PRE(nodeSize >= (at - ep_name));
+	memcpy(node, ep_name, ((at - ep_name)-1));
+	return 0;
+}
+
+/**
+ * Used to decode the ip and port from the given end point
+ * ep_name : endpoint address from domain
+ * node    : copy ip address from ep_name
+ * port    : copy port number from ep_name
+ * Example of ep_name IPV4 192.168.0.1:4235
+ *                    IPV6 [4002:db1::1]:4235
+ */
+static int libfab_ep_addr_decode_native(const char *ep_name, char *node,
+					size_t nodeSize, char *port,
+					size_t portSize)
+{
+	char   *name;
+	char   *cp;
+	size_t  n;
+	int     rc = 0;
+
+	M0_PRE(ep_name != NULL);
+
+	M0_ENTRY("ep_name=%s", ep_name);
+
+	name = (char *)ep_name + strlen("libfab:");
+
+	if( name[0] == '[' ) {
+		/* IPV6 pattern */
+		cp = strchr(name, ']');
+		if (cp == NULL)
+			return M0_ERR(-EINVAL);
+
+		name++;
+		n = cp - name;
+		if (n == 0 )
+			return M0_ERR(-EINVAL);
+		cp++;
+		if (*cp != ':')
+		return M0_ERR(-EINVAL);
+		cp++;
+	}
+	else {
+		/* IPV4 pattern */
+		cp = strchr(name, ':');
+		if (cp == NULL)
+			return M0_ERR(-EINVAL);
+
+		n = cp - name;
+		if (n == 0 )
+			return M0_ERR(-EINVAL);
+
+		++cp;
+	}
+
+	M0_PRE(nodeSize >= (n+1));
+	M0_PRE(portSize >= (strlen(cp)+1));
```

Comment:
```
M0_ASSERT(nodeSize >= (n + 1));
M0_ASSERT(portSize >= (strlen(cp) + 1));
```

It's better to handle error cases here, instead of panic.

Reply:
Handled error cases and removed asserts